docs/adr/0001-mvp-webidl.md

```diff
@@ -0,0 +1,179 @@
+* Good, because the `weedle` crate provides a ready-made parser for WebIDL.
+* Good, because WebIDL has some base level of familiarity around Mozilla.
+* Bad, because developers will need to duplicate their API, once in the Rust code and once in the IDL.
+* Bad, because WebIDL is designed for a different use-case, so it's likely to be an awkward fit.
+* Bad, because `weedle` doesn't generate particularly helpful error messages (it seems designed for
+  parsing known-good WebIDL definitions rather than helping you develop new ones).
+
+Ultimately, this seems like the lowest-cost way to get started, while deferring the
+important-but-not-existentially-risky work of making an IDL experience that fits really
+well with Rust code.
+
+### Option B: Use an external interface definition file based on a custom language.
+
+We can require developers to specify their component API in an external definition file,
+using our own custom variant of an IDL syntax.
+
+* Good, because the syntax can be custom designed to fit well with the developer's mental
+  model of the generated code.
+* Good, because we already have several different IDL variants in use at Mozilla (WebIDL, XPIDL),
+  so our changes of building somethat that feels familiar are high.
+* Bad, because developers will need to duplicate their API, once in the Rust code and once in the IDL.
+* Bad, because we have to make up and document a whole syntax.
+* Bad, because rust parsing crates such as `nom` do not seem to generate particularly helpful error messages
+  by default, adding friction for the developer.
+
+Ultimately, while a custom syntax would probably "feel" better from the consumer's perspective
+than WebIDL, the costs involved are not worth that tradeoff for the MVP. Beyond the MVP we expect
+that direct annotation of the Rust code will provide a better developer experience, leaving this
+option as an unnecessary middle-ground.
+
+### Option C: Infer the API directly from the rust code using annotations and macros.
+
+We can allow developers to sprinkle some macro annotations directly on their rust code
+in order to declare the component API, similar to the approach taken by `wasm-bindgen`.
+
+* Good, the Rust code is a single source of truth for the API definition.
+* Good, because developers are familiar with this approach from other tools.
+* Bad, because our team doesn't have much experience working with macros at scale.
+* Bad, because from a poke around in the `wasm-bindgen` code, they seem to need to do
+  some pretty scary things in order to make the macros Just Work in various edge-cases.
+
+Ultimately, this feels like a good approach longer-term, but risks being too much of a time-sink
+for the MVP.
+
+### Option D: Use build-scripts and macros to deeply integrate with the rust crate's build system.
+
+We can encourage consumers to structure their rust component as a crate, take a build-time dependency
+on our tool, and magic things into existence as part of `cargo build`.
```

Comment:
@eoger, is this kinda the approach you're using in the FxA client demo? I know `uniffi_bindgen` doesn't have an elaborate API for calling it from `build.rs`, but would you say it's a hybrid between D and E?